### ISSUE TITLE
vlm_mtp decode silently drops per-request logits processors (grammar, thinking_budget, penalties)

### ISSUE THREAD

**Issue by picosquare** [jundot/omlx#2399]
Disclaimer: This bug was found while coding on a different project that uses omlx + structured_output. The report was written by Claude Code. I manually verified the witness script below and can confirm the issue.

**Describe the bug**

When `vlm_mtp_enabled` is on, structured-output requests come back **unconstrained**: HTTP 200, no `Warning` header, nothing in the log. The grammar *is* compiled — a malformed schema is rejected in under 40 ms with xgrammar's own error text — but the compiled mask never reaches the sampler on the vlm_mtp decode path.

Because `_route_to_vlm_mtp` serves only **one** request at a time, the bug is intermittent in a very specific way: requests that fall back to BatchGenerator *are* correctly constrained. So the same body gives 0/6 constrained when sent sequentially and 4/6 when sent concurrently. That contrast is the clearest signal — the mask exists and works, it is just skipped on the fast path.

`thinking_budget` and the repetition / presence / frequency penalties are dropped on the same path, for the same reason.

**To Reproduce**

1. Load a model with a vlm_mtp drafter attached — here `gemma-4-26B-A4B-it-qat-4bit` (`engine_type: vlm`) with `gemma-4-26B-A4B-it-qat-assistant-4bit`. Confirm via `/v1/models/status`.
2. Make sure `vlm_mtp_enabled` is on for that model.
3. Save the script below as `repro.sh`, set `OMLX_URL` and `OMLX_API_KEY`, run `bash repro.sh`.
4. Compare the two tallies.

```bash
#!/usr/bin/env bash
# The grammar `root ::= "red"` admits exactly the three characters `red`.
# Anything else — "Red.", a sentence — proves no mask was applied.
set -eu

URL="${OMLX_URL:-http://localhost:8000}/v1/chat/completions"
MODEL="${OMLX_MODEL:-gemma-4-26B-A4B-it-qat-4bit}"
N="${N:-6}"

body=$(cat <<JSON
{
  "model": "$MODEL",
  "messages": [{"role": "user", "content": "Name a primary colour."}],
  "max_tokens": 32,
  "temperature": 0,
  "structured_outputs": {"choice": ["red"]}
}
JSON
)

tmp=$(mktemp -d); trap 'rm -rf "$tmp"' EXIT

ask() { curl -sS "$URL" -H 'Content-Type: application/json' \
        -H "Authorization: Bearer $OMLX_API_KEY" -d "$body" > "$1"; }

tally() {
  local ok=0 out mark
  for i in $(seq 1 "$2"); do
    out=$(jq -r '.choices[0].message.content // .error.message // "(no answer)"' "$tmp/$i.json")
    if [ "$out" = "red" ]; then ok=$((ok+1)); mark="constrained  "
    else                        mark="UNCONSTRAINED"; fi
    printf '  %d  %s  %s\n' "$i" "$mark" "$out"
  done
  printf '%s: %d/%d constrained\n\n' "$1" "$ok" "$2"
}

echo "== $N sequential =="
for i in $(seq 1 "$N"); do ask "$tmp/$i.json"; done
tally "sequential" "$N"

echo "== $N concurrent =="
for i in $(seq 1 "$N"); do ask "$tmp/$i.json" & done
wait
tally "concurrent" "$N"
```

Actual output:

```
== 6 sequential ==
  1  UNCONSTRAINED  Red.
  2  UNCONSTRAINED  Red.
  3  UNCONSTRAINED  Red.
  4  UNCONSTRAINED  Red.
  5  UNCONSTRAINED  Red.
  6  UNCONSTRAINED  Red.
sequential: 0/6 constrained

== 6 concurrent ==
  1  constrained    red
  2  constrained    red
  3  UNCONSTRAINED  Red.
  4  UNCONSTRAINED  Red.
  5  constrained    red
  6  constrained    red
concurrent: 4/6 constrained
```

If VLM MTP is disabled, the output is as expected:

```
== 6 sequential ==
  1  constrained    red
  2  constrained    red
  3  constrained    red
  4  constrained    red
  5  constrained    red
  6  constrained    red
sequential: 6/6 constrained

== 6 concurrent ==
  1  constrained    red
  2  constrained    red
  3  constrained    red
  4  constrained    red
  5  constrained    red
  6  constrained    red
concurrent: 6/6 constrained
```

A `{"json": {"const": "red"}}` schema behaves identically, as does `response_format` with a strict `json_schema`.

**Expected behavior**

All twelve answers in the first block should be exactly red — as the second block shows they are once vlm_mtp is off. A grammar `root ::= "red"` cannot produce a capital letter or a full stop, so every `Red.` above is a request in which the compiled mask was not applied.

Failing that, the request should **fail loudly** — the same way the diffusion path already does with HTTP 400 — rather than returning free-form text as if the schema had been honoured. Silently returning unconstrained output is the worst of the three outcomes, because a caller cannot tell it from a model that simply chose to answer in the requested shape.

**Screenshots**

Not applicable — the script output above is the evidence.

**Desktop (please complete the following information):**

- macOS Version: 26.5.2
- oMLX Version: 0.5.3 (`31d07a7`); the same code is present on `main` (`b8a272c`, 2026-07-29)
- Hardware: Apple Silicon
- Model: `gemma-4-26B-A4B-it-qat-4bit` (`engine_type: vlm`) with `gemma-4-26B-A4B-it-qat-assistant-4bit` as the vlm_mtp drafter. Same behaviour on `gemma-4-31B-it-qat-4bit`.

**Additional context**

*Root cause (read from the source, not inferred from the symptom):*

`omlx/scheduler.py:4571-4581` builds the `GrammarConstraintProcessor` and puts it into `logits_processors`. On the BatchGenerator path those are handed to `insert(..., logits_processors=[per_row_lps])` (`:8522-8538`, `:4260-4275`) and applied in mlx-lm `generate.py:1346`.

But `omlx/scheduler.py:8488-8491` routes eligible requests to `_route_to_vlm_mtp(request, cache, tokens, sampler, sm)` — and that function (`:6577-6584`) never receives `logits_processors`. Only the suppress-token processor is carried over, by hand, via `_make_suppressing_sampler` at `:6643`. `run_vlm_mtp_decode` (`omlx/speculative/vlm_mtp.py:306-321`), `_VLMMTPDecodeState` (`scheduler.py:120-141`) and `_step_vlm_mtp` (`:6787ff`) have no notion of logits processors at all.

So grammar constraints, `thinking_budget` and the sampling penalties are silently discarded on this path. On `main` the same call is made from `:4600-4602` and `:9036-9038`. The one-at-a-time behaviour at `:6615-6622` ("drafter is busy … falling back to BatchGenerator") is what produces the sequential/concurrent split above.

*Why this reads as an oversight rather than a documented limitation:*

- The diffusion path **explicitly rejects** grammar with HTTP 400 (`omlx/engine/vlm.py:3570-3574`), and DFlash effectively does the same (#1586). oMLX's own convention here is to refuse, not to ignore.
- #1241 / #1564 added a `Warning` header precisely so that an unenforced `response_format` is never silent. It does not fire here, because compilation *succeeded* — the schema is dropped afterwards.
- The `vlm_mtp_enabled` docstring (`omlx/model_settings.py:202-206`) mentions only the BatchGenerator bypass and exclusivity with other speculative paths. Structured output is not mentioned.
- `tests/test_grammar_live.py` covers Qwen3.5-4B, gemma-3-4b and gpt-oss — all on the `batched` engine. No vlm_mtp model exercises this path.
- This is the same failure class as #1823 ("Grammar and thinking_budget silently stop applying under concurrent mixed load"), fixed by #1824 / #1845 for BatchGenerator. vlm_mtp appears to have been missed in that pass.

*Suggested fix:*

Thread the per-request `logits_processors` into `_route_to_vlm_mtp` / `_VLMMTPDecodeState` and apply them to the target logits before sampling in `run_vlm_mtp_decode`'s round loop, and before each drafted-token verify. If applying a grammar mask during speculative verification is not practical, please fail loudly instead — reject with HTTP 400 as the diffusion path does, or set the #1564 `Warning` header.

*Workaround:* disabling `vlm_mtp_enabled` for the affected models restores enforcement, at the cost of the MTP speedup.

**Comment by jundot** [comment]
Thanks for the detailed report, the analysis was exactly right. The vlm_mtp wrapper hands mlx-vlm's round loop a bare sampler, so grammar masks, thinking budget and the penalties had no application point on that path.

I did look at threading the processors into the loop the way you suggested, and I don't think it is realistic. The round loop lives in upstream mlx-vlm and only exposes a sampler that sees logits with no token history, while these processors are stateful: penalties and thinking budget need the full generated sequence, and a grammar matcher has to advance on every accepted token and roll back whenever a draft is rejected mid-verify. Supporting that would mean patching mlx-vlm's speculative internals, and even then grammar would gain very little, because the drafter keeps proposing unconstrained tokens that the masked verify rejects, so acceptance collapses and the MTP speedup mostly disappears with it. The native MTP path handles grammar the same way for the same reason, it steps aside rather than applying the mask inside the verify cycle.

So as of 9387ebdd on main, a request carrying any of these parameters falls back to BatchGenerator where everything is enforced, instead of coming back unconstrained. Model settings that would force that fallback on every request (guided grammar, thinking budget, penalty defaults) can no longer be combined with vlm_mtp_enabled, and both settings UIs explain the conflict. You keep the MTP speedup for plain requests and correct output for constrained ones, and nothing is silent anymore. This will be in the next release.
